yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,412 @@
+        (tgt_len x batch_size x src_len x vocab_size), and the
+        transition tensor returned is of shape
+        (tgt_len x batch_size x src_len x src_len).
+
+        Args:
+            batch (data.PaddedBatch).
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: tensosr of emission
+                 and transition probabilities for each transition
+                 state (i.e., target symbol).
+        """
+        encoder_out = self.source_encoder(batch.source).output
+        if self.has_features_encoder:
+            encoder_features_out = self.features_encoder(batch.features).output
+            # Averages to flatten embedding.
+            encoder_features_out = encoder_features_out.sum(
+                dim=1, keepdim=True
+            )
+            # Expands length of source.
+            encoder_features_out = encoder_features_out.expand(
+                -1, encoder_out.shape[1], -1
+            )
+            # Concatenates with average.
+            encoder_out = torch.cat(
+                [encoder_out, encoder_features_out], dim=-1
+            )
+        if self.training:
+            output = self.decode(
+                encoder_out,
+                batch.source.mask,
+                batch.target.padded,
+            )
+        else:
+            output = self.greedy_decode(
+                encoder_out,
+                batch.source.mask,
+            )
+        return output
+
+    def decode(
+        self,
+        encoder_out: torch.Tensor,
+        encoder_mask: torch.Tensor,
+        target: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Decodes a sequence given the encoded input.
+
+        Decodes until all sequences in a batch have reached [EOS] up to
+        length of `target` args.
+
+        Args:
+                encoder_out (torch.Tensor): batch of encoded input symbols.
+                    Shape: (batch_size x src_len x encoder_hidden*num_dir).
+                encoder_mask (torch.Tensor): mask for the batch of encoded
+                    input symbols. Shape: (batch_size x src_len).
+                target (torch.Tensor): target symbols. Decodes up to
+                    `len(target)` symbols.
+
+        Returns:
+                all_log_probs: (torch.Tensor): tensor of emission
+                    probabilities for each transition state (target symb).
+                    Shape: (tgt_len x batch_size x src_len x vocab_size)
+                all_alignements: (torch.Tensor): tensor of transition
+                    probabilities for each transition state (target symb).
+                    Shape: (tgt_len x batch_size x src_len x src_len)
+        """
+        log_probs, alignment_matrix, decoder_hiddens = self.init_decoding(
+            encoder_out, encoder_mask
+        )
+        all_log_probs, all_alignments = [log_probs], [alignment_matrix]
+        for t in range(target.size(1)):
+            tgt_symbol = target[:, t]
```

Comment:
I think we need to address the `teacher_forcing` flag somehow. If it does not make sense to turn off teacher forcing in this architecture, maybe we should log a warning in case it is turned off?

Reply:
Do you want a warning or straight assertion failure?